chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,378 @@
+        self.delta_d = delta_d
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.max_n_clusters = max_n_clusters
+
+        # L1 or L2 norm is allowed only
+        if norm == 1:
+            self.norm = self._l1_norm
+        elif norm == 2:
+            self.norm = self._l2_norm
+        else:
+            raise ValueError("Norm can only be 1 or 2")
+
+    def _l1_norm(self, x, axis):
+        """Function to calculate L1 Norm by given axes
+
+        Args:
+            x (nd-array): Input matrix to calculate norm
+            axis (int / tuple): Axes information for norm calculation
+
+        Returns:
+            nd-array : Norm applied to given axes
+        """
+        return sum(absolute(x), axis=axis)
+
+    def _l2_norm(self, x, axis):
+        """Function to calculate L2 Norm by given axes
+
+        Args:
+            x (nd-array): Input matrix to calculate norm
+            axis (int / tuple): Axes information for norm calculation
+
+        Returns:
+            nd-array : Norm applied to given axes
+        """
+        return sum(x ** 2, axis=axis)
+
+    def _variance_term(self, pred, gt, means, delta_v, gt_idx):
+        """Function to calculate variance term
+
+        Args:
+            pred (nd-array): Prediction output
+            gt (nd-array): Ground truth output
+            means (nd-array): Instance means
+            delta_v (float): Coefficient to decide 'pull force' power
+            gt_idx (tuple / nd-array): Indexes of ground truth instances
+        Returns:
+            float : variance loss
+        """
+
+        bs, n_filters, n_loc = pred.shape
+        n_instances = gt.shape[1]
+
+        # Prepare each item with same size by broadcasting missing axes
+        p = expand_dims(pred, axis=1)
+        g = expand_dims(gt, axis=2)
+        m = expand_dims(means, axis=3)
+
+        p = broadcast_to(p, (bs, n_instances, n_filters, n_loc))
+        g = broadcast_to(g, (bs, n_instances, n_filters, n_loc))
+        m = broadcast_to(m, (bs, n_instances, n_filters, n_loc))
+
+        m = cast(m, p.dtype)
+        g = cast(g, p.dtype)
+
+        module = cuda.get_array_module(p)
+        dv = module.asarray(delta_v, p.dtype)
+
+        _var = self.norm((p - m), 2)
+        _var = maximum(module.asarray(0.0, p.dtype),
+                       _var - dv) ** 2  # Suppress inlier distance
+        _var = _var * g[:, :, 0, :]
+
+        var_term = 0.0
+        for i in range(bs):
+            var_term += sum(_var[i, gt_idx[i]]) / sum(g[i, gt_idx[i], 0])
+        var_term /= bs
+
+        return var_term
+
+    def _distance_term(self, means, delta_d, n_objects):
+        """Function to calculate distance term
+
+        Args:
+            means (nd-array): Instance means
+            delta_d (float): Coefficient to decide 'push force' power
+            n_objects (nd-array): Instance count in current input
+
+        Returns:
+            float : distance loss
```

Comment:
ditto (the returned value is Chainer variable)

Reply:
Please address the above comment.